daras_ai_v2/base.py

```diff
@@ -1594,23 +1647,57 @@ def render_extra_waiting_output(self):
     def estimate_run_duration(self) -> int | None:
         pass
 
-    def submit_and_redirect(self):
+    def submit_and_redirect(self) -> typing.NoReturn | None:
         sr = self.on_submit()
         if not sr:
             return
         raise gui.RedirectException(self.app_url(run_id=sr.run_id, uid=sr.uid))
 
+    def publish_and_redirect(self) -> typing.NoReturn | None:
+        assert self.request.user and not self.request.user.is_anonymous
+
+        if (
+            self.get_run_state(gui.session_state) == RecipeRunState.starting
+            and self.current_sr_user
+        ):
+            with transaction.atomic():
+                # select_for_update() is important because we need to ensure the
+                # run is not being modified by another thread (e.g. in a different
+                # tab). otherwise, we could end up queueing the run twice.
+                sr = SavedRun.objects.select_for_update().get(id=self.current_sr.id)
+                if self.get_run_state(sr.to_dict()) != RecipeRunState.starting:
+                    # run has already been started in another tab
+                    gui.rerun()
+                if (
+                    sr.uid == self.current_sr_user.uid
+                    and self.current_sr_user.is_anonymous
+                ):
+                    # run belonged to an anonymous user who now logged
+                    # in from an existing account now, so we update the uid
+                    sr.uid = self.request.user.uid
+                    sr.save(update_fields=["uid"])
```

Comment:
PS didn't need to save here as `dump_state_to_sr` would have done it  anyways

Reply:
i'm aware, but i added this because it isn't obvious (also would not be obvious to fix if we were to later refactor `dump_state_to_sr` to only save the state and not everything else)